### ISSUE TITLE
"Synchronous operations are disallowed."  in ViewExecutor.ExecuteAsync() when calling async tag helpers without awaiting

### ISSUE THREAD

**Issue by neoGeneva** [aspnet/AspNetCore#14685]
It's possible to use an async tag helper in a way that doesn't generate any compiler errors or warnings, and causes a "Synchronous operations are disallowed" exception.

I understand that this is because I'm doing something weird, but I expect there's other people doing the same thing  and will have the same problem, and it's not exactly obvious why the errors occurring, so it would be nice if there was a clearer error thrown or, ideally, a compiler error.

So for this to happen there has to be a couple of things to be true:

1. The page has to be long enough before using a form tag
2. The user can't have a antiforgery cookie
3. The form tag helper mustn't be awaited correctly

The problem is caused by the following cshtml, which I use because the `@helper` syntax was removed in aspnet core 1 (and googling "aspnet core helper directive" this syntax is often suggested), and prior to 3 there was no alternative:
```
@{
    Func<object, IHtmlContent> ShowForm = @<form asp-action="Home" method="post"></form>;
}

@ShowForm(null)
```

The stacktrace for the "Synchronous operations are disallowed" exception:
```
System.InvalidOperationException: Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead.
   at Microsoft.AspNetCore.Server.IIS.Core.HttpResponseStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at Microsoft.AspNetCore.Server.IIS.Core.WrappingStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.FlushInternal(Boolean flushEncoder)
   at Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ActionContext actionContext, IView view, ViewDataDictionary viewData, ITempDataDictionary tempData, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.ExecuteAsync(ActionContext context, ViewResult result)
   at Microsoft.AspNetCore.Mvc.ViewResult.ExecuteResultAsync(ActionContext context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeResultAsync>g__Logged|21_0(ResourceInvoker invoker, IActionResult result)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|29_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeResultFilters>g__Awaited|27_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at RubberMonkey.Sales.Web.LookupNameRewriteMiddleware.Invoke(HttpContext context) in C:\Projects\rm\Main\Source\RubberMonkey.Sales.Web\LookupNameRewriteMiddleware.cs:line 95
   at RubberMonkey.Sales.Web.AntibotMiddleware.Invoke(HttpContext httpContext) in C:\Projects\rm\Main\Source\RubberMonkey.Sales.Web\Middleware\AntibotMiddleware.cs:line 65
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at RubberMonkey.Sales.Web.CustomExceptionHandlerMiddleware.Invoke(HttpContext context) in C:\Projects\rm\Main\Source\RubberMonkey.Sales.Web\Middleware\CustomExceptionHandlerMiddleware.cs:line 33
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.HandleException(HttpContext context, ExceptionDispatchInfo edi)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()
```

And this is the stacktrace for a "Headers are read-only, response has already started" exception (when debugging locally), which is thrown immediately before the async one:
```
System.InvalidOperationException: Headers are read-only, response has already started.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ThrowHeadersReadOnlyException()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.Microsoft.AspNetCore.Http.IHeaderDictionary.set_Item(String key, StringValues value)
   at Microsoft.AspNetCore.Http.ResponseCookies.Append(String key, String value, CookieOptions options)
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgeryTokenStore.SaveCookieToken(HttpContext httpContext, String token)
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgery.SaveCookieTokenAndHeader(HttpContext httpContext, String cookieToken)
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgery.GetAndStoreTokens(HttpContext httpContext)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.AntiforgeryExtensions.GetHtml(IAntiforgery antiforgery, HttpContext httpContext)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.DefaultHtmlGenerator.GenerateAntiforgery(ViewContext viewContext)
   at Microsoft.AspNetCore.Mvc.TagHelpers.FormTagHelper.Process(TagHelperContext context, TagHelperOutput output)
   at Microsoft.AspNetCore.Razor.TagHelpers.TagHelper.ProcessAsync(TagHelperContext context, TagHelperOutput output)
   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.RunAsync(TagHelperExecutionContext executionContext)
   at AspNetCore.Areas_Admin_Views_Orders_Payment.<>c__DisplayClass17_1.<<ExecuteAsync>b__7>d.MoveNext() in C:\Projects\rm\Main\Source\RubberMonkey.Sales.Web\Areas\Admin\Views\Orders\Payment.cshtml:line 89
```

### To Reproduce
1. Create an aspnet core mvc project
2. Replace the code in `index.cshtml` with the code below
3. Start the site and load the page
4. Remove the antiforgery cookie in your browser and reload 

Index.cshtml
``` cshtml
@using Microsoft.AspNetCore.Html
@{
    ViewData["Title"] = "Home Page";

    Func<object, IHtmlContent> ShowForm = @<form asp-action="Home" method="post"></form>;
}

@for (var i = 0; i < 10000; ++i)
{
    <div style="display:none">this is some content</div>
}

@ShowForm(null)
```

And, I can prevent the error by using the syntax introduced in 3:

``` cshtml
@{
    ViewData["Title"] = "Home Page";

    async Task ShowTable2()
    {
        <form asp-action="Home" method="post"></form>
    }
}

@for (var i = 0; i < 10000; ++i)
{
    <div style="display:none">this is some content</div>
}

@{
    await ShowTable2();
}
```

### Additional context
```
> dotnet --info

  It was not possible to find any installed .NET Core SDKs
  Did you mean to run .NET Core SDK commands? Install a .NET Core SDK from:
      https://aka.ms/dotnet-download

Host (useful for support):
  Version: 3.0.0
  Commit:  7d57652f33

.NET Core SDKs installed:
  No SDKs were found.

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
The scenario you're describing is not supported. When calling async TagHelpers the calls must be awaited.

**Comment by neoGeneva** [comment]
Hey thanks @mkArtakMSFT 

I guess the real issue here isn't that the syntax I used isn't supported, it's that it's really hard to tell that's the case, for a couple of reasons:

1. There's nothing in the editor (in this case vs2019), or compiler to show the syntax isn't supported
2. The syntax works fine, unless the TagHelper is async, and there's nothing to show that TagHelpers are async
3. The runtime error generated is very disconnected from the cause of the issue

For the new syntax you get a an error saying "MVC1006: This method contains a TagHelper and therefore must be async to return a task...", which is great, and it'd be good if there was something similar for the syntax I used (or an error outright disallowing the syntax)

That said, I appreciate it's a rare and minor issue, so I don't really expect anything to be fixed here, though maybe if someone else has the same problem they'll see this issue and know how to fix it themselves.
